### ISSUE TITLE
BUG: Opening files with spaces fails

### ISSUE THREAD

**Issue by cjermain** [jupyter/notebook#728]
Just installed the latest Jupyter on Python 3.4.1 with:
```bash
$ pip install jupyter
$ pip freeze | grep jupyter
jupyter==1.0.0
jupyter-client==4.1.1
jupyter-console==4.0.3
jupyter-core==4.0.6
```
I'm having an issue opening files that contain spaces in their names. The notebook launches, but it tires to open two notebooks instead of a single notebook.
```
$ jupyter notebook Awesome\ Notebook.ipynb                                      
[I 10:48:28.080 NotebookApp] The port 8888 is already in use, trying another random port.
[I 10:48:28.085 NotebookApp] Serving notebooks from local directory: /home/user/
[I 10:48:28.085 NotebookApp] 0 active kernels 
[I 10:48:28.085 NotebookApp] The IPython Notebook is running at: http://localhost:8889/
[I 10:48:28.085 NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[W 10:48:28.671 NotebookApp] 404 GET /notebooks/Awesome (::1): No such file or directory: Awesome
[W 10:48:28.717 NotebookApp] 404 GET /notebooks/Awesome (::1) 48.08ms referer=None
```
This also occurs with:
```
$ jupyter notebook "Awesome Notebook.ipynb"
```
Both should work.

I think the bug is in `parse_command_line` in [notebookapp.py](https://github.com/jupyter/notebook/blob/4578c34b0f999735ee49e1492be3dd5941951551/notebook/notebookapp.py#L788), but haven't had a chance to dig deeper yet. In previous IPython versions I have not had this issue, so it seems like a regression.

Are there tests for this issue?

**Comment by Carreau** [comment]
bash I suppose ? Linux ? Os X ? which version of `traitlets` ?

It's working fine here with the above version you list.

**Comment by cjermain** [comment]
I'm using ZSH on OpenSUSE 13.2. I also just tried bash with no success either.
```bash
$ pip freeze | grep traitlets
traitlets==4.0.0
```

**Comment by cjermain** [comment]
I can confirm that the system arguments going into `jupyter notebook` are correctly quoted and grouped (I just printed `sys.argv` in the `jupyter` bash script), so I don't think its a OS issue. I also tested putting in two files at the same time, which gives the correct `sys.argv`.

**Comment by cjermain** [comment]
Thanks for checking. I have the same problem on Python 2. I've checked manually down to [notebookapp.py#L806](https://github.com/jupyter/notebook/blob/4578c34b0f999735ee49e1492be3dd5941951551/notebook/notebookapp.py#L806), and everything looks fine if I print the Notebook Config:
```
{'NotebookApp': {'file_to_run': '/home/user/Awesome Notebook.ipynb'}}
```
However, I'm still having the issue. I'll keep digging manually to see if I can isolate the issue.

**Comment by cjermain** [comment]
Could this be in the URL generation or interaction with my browser (Firefox 42.0)? If I go to the correct URL, everything is fine.

**Comment by Carreau** [comment]
> Could this be in the URL generation or interaction with my browser (Firefox 42.0)? If I go to the correct URL, everything is fine.

might be. In `notebookapp.py` around line 1100 should be a:

```
b = lambda : browser.open(url_path_join(self.connection_url, uri),
                                          new=2)
```

you can try to add debug statement around that. We might need to encode the URI there.

**Comment by cjermain** [comment]
Ok, here is the issue ([notebookapp.py#L1101](https://github.com/jupyter/notebook/blob/4578c34b0f999735ee49e1492be3dd5941951551/notebook/notebookapp.py#L1101)).
```python
>> print(url_path_join(self.connection_url, uri))
http://localhost:8890/notebooks/Awesome Notebook.ipynb
```
Notice there is no `%20` for the space. This is not a valid URL. There should be a line like:
```python
uri.replace('' ", "%20")
```
This solves the issue. Do you agree? I'll put a pull request if so.

**Comment by Carreau** [comment]
you don't want a replace you likely want to `urllib.quote(url)` as things like, `\`,`:` and other have to be encoded too. 

But otherwise, yes, PR welcome.

**Comment by willingc** [comment]
@Carreau @cjermain Here's an example from the nbgrader project: https://github.com/jupyter/nbgrader/commit/0bd45410a3851f5ca0dad9b870b1fab02681a43e

**Comment by Carreau** [comment]
> jupyter/nbgrader@0bd4541

the escaping there is in js, here the escape as likely to be in Python.

**Comment by willingc** [comment]
@Carreau Yep. Passed it along for more of an example of all the characters. Looks like @cjermain already opened a PR.

**Comment by cjermain** [comment]
Looks like we already have a `url_escape` method which uses `urllib.quote` ([utils.py#L62](https://github.com/jupyter/notebook/blob/master/notebook/utils.py#L62)), with tests ([test_utils.py#L24](https://github.com/jupyter/notebook/blob/master/notebook/tests/test_utils.py#L24)). It looks like its no used in `notebookapp.py`, which is weird. I'll change the PR to use `url_escape`.

**Comment by cjermain** [comment]
Thanks!

**Comment by Carreau** [comment]
Thanks to you for finding, reporting, debugging and patching ! You did most of the work !
